scatter plot and hexbin plot lose x-axis when colorbar is included.

javadnoorb opened the pull request:
closes https://github.com/pandas-dev/pandas/issues/10611 
closes https://github.com/pandas-dev/pandas/issues/10678. 

The x-axis for scatter plot and hexbin plot disappears when colorbar is included. This seems to be because colorbar axis is looped through in `_handle_shared_axes`: https://github.com/pandas-dev/pandas/blob/8a5830305d2d9a087fbe46bd968218104ffdfc49/pandas/plotting/_core.py#L426
after discussing with @TomAugspurger (issue https://github.com/pandas-dev/pandas/issues/10611), we decided to try adding the attribute `__is_pandas_colorbar` to colorbar axis object and skipping it during handling of shared axes. 

I've done some tests that seem to fix the issue. But we may need more tests:
```
%matplotlib inline
import matplotlib.pylab as pl
from mypandas import pandas as pd
import numpy as np
random_array = np.random.random((1000,3))

df = pd.DataFrame(random_array,columns=['A label','B label','C label'])
df.plot.scatter('A label','B label',c='C label', PATCH_MODE_FLAG = False);pl.title('pandas current version');
df.plot.scatter('A label','B label',c='C label', PATCH_MODE_FLAG = True);pl.title('patch fixing x-axis');

df.plot.hexbin('A label','B label', gridsize=25, PATCH_MODE_FLAG = False);pl.title('pandas current version');
df.plot.hexbin('A label','B label', gridsize=25, PATCH_MODE_FLAG = True);pl.title('patch fixing x-axis');
```

![image](https://user-images.githubusercontent.com/26352146/37747996-49642266-2d58-11e8-8c6f-c382267e79db.png)
![image](https://user-images.githubusercontent.com/26352146/37747999-4d88d986-2d58-11e8-8445-f1348c35d892.png)
![image](https://user-images.githubusercontent.com/26352146/37748001-5109d93e-2d58-11e8-9f6f-73a68329f731.png)
![image](https://user-images.githubusercontent.com/26352146/37748005-54beb590-2d58-11e8-8233-4e26b135fdad.png)

jorisvandenbossche committed:
DOC: update vendored numpydoc version

Numpydoc upstream at 21a194e9b42ebe95a70475e35cb69bcb4801ff36

jorisvandenbossche committed:
DOC: use numpydoc_use_blockquotes=False for compatibility

jorisvandenbossche committed:
DOC: change to numpydoc: add option to use member listing for attributes

javadnoorb committed:
removed colorbars from _handle_shared_axes when called by scatterplot and hexbin

javadnoorb committed:
removed colorbars from _handle_shared_axes when called by scatterplot and hexbin

javadnoorb committed:
added a debug global variable

l736x committed:
DOC: Improve the docstrings of CategoricalIndex.map (#20286)

shelvinskyi committed:
DOC: update the DataFrame.to_hdf() docstirng (#20186)

jorisvandenbossche committed:
DOC: enable docstring on DataFrame.columns/index (#20385)

jorisvandenbossche committed:
DOC: general docstring formatting fixes (#20449)

minggli committed:
parameterize tests in scalar/timedelta (#20428)

TomAugspurger commented:
Thanks, let us know when you're ready for us to take a look.

How does this work with `subplots=True`? Do we not have to worry about that since it's a scatter plot?

How does this work with user-specified axes via `ax=`? Especially with multiple subplots. It may already be broken, so no worries if it doesn't work well.

javadnoorb commented:
Yep, it's already broke. Please feel free to take a look:

```
def make_multiplots(numrows=2,numcols=2,sharey=True,sharex=True,PATCH_MODE_FLAG = True, is_scatter = True):
    fig, axes = pl.subplots(numrows,numcols,sharey=sharey,sharex=sharex )
    fig.set
    for i in range(axes.shape[0]):
        for j in range(axes.shape[1]):
            random_array = np.random.random((1000,3))
            df = pd.DataFrame(random_array,columns=['A label','B label','C label'])
            if is_scatter:
                df.plot.scatter('A label','B label',c='C label', PATCH_MODE_FLAG = PATCH_MODE_FLAG,ax=axes[i,j]);
            else:
                df.plot.hexbin('A label','B label',gridsize=20, PATCH_MODE_FLAG = PATCH_MODE_FLAG,ax=axes[i,j]);
            
            
make_multiplots(numrows=2,numcols=2,sharey=True, sharex=True, PATCH_MODE_FLAG = False, is_scatter=True);pl.suptitle('pandas current version');            
make_multiplots(numrows=2,numcols=2,sharey=True, sharex=True, PATCH_MODE_FLAG = True, is_scatter=True);pl.suptitle('patch fixing x-axis');            
make_multiplots(numrows=2,numcols=2,sharey=True, sharex=True, PATCH_MODE_FLAG = False, is_scatter=False);pl.suptitle('pandas current version');            
make_multiplots(numrows=2,numcols=2,sharey=True, sharex=True, PATCH_MODE_FLAG = True, is_scatter=False);pl.suptitle('patch fixing x-axis');  
```          
![image](https://user-images.githubusercontent.com/26352146/37773292-97f0a5a2-2db3-11e8-9912-8f2f558f5740.png)
![image](https://user-images.githubusercontent.com/26352146/37773303-9eaa8e08-2db3-11e8-975b-ce454a33e277.png)
![image](https://user-images.githubusercontent.com/26352146/37773312-a30c9b62-2db3-11e8-9040-864b16ebf59a.png)
![image](https://user-images.githubusercontent.com/26352146/37773316-a6259916-2db3-11e8-9be2-592295972768.png)

TomAugspurger commented on pandas/plotting/_core.py:
```diff
@@ -870,6 +870,8 @@ def _make_plot(self):
         scatter = ax.scatter(data[x].values, data[y].values, c=c_values,
                              label=label, cmap=cmap, **self.kwds)
         if cb:
+            if PATCH_MODE:
```
I assume this ``PATCH_MODE` is just for debugging?

javadnoorb replied:
yes, it should be removed. Didn't know of a better way to do this.

TomAugspurger commented on pandas/plotting/_tools.py:
```diff
@@ -298,20 +298,20 @@ def _remove_labels_from_axis(axis):
 def _handle_shared_axes(axarr, nplots, naxes, nrows, ncols, sharex, sharey):
     if nplots > 1:
-
+        
```
This whitespace will likely fail the linter

TomAugspurger commented on pandas/plotting/_tools.py:
```diff
@@ -298,20 +298,20 @@ def _remove_labels_from_axis(axis):
                 for ax in axarr:
                     layout[ax.rowNum, ax.colNum] = ax.get_visible()
-
+                    
```
Likewise

TomAugspurger reviewed (commented):
I would say make a new issue for subplots, and ignore it for now.

As for testing, how about we make two plots, one with `c` and one without, and assert that the tick labels are the same?


```
fig, ax = plt.supblots()
df.plot.scatter(x='x', y='y', ax=ax)

fig, ax2 = plt.subplots()
df.plot.scatter(x='x', y='y', c='c', ax=ax2)

assert ax.get_xticklabels() == ax2.get_xticklabels()
```

or something like that.

javadnoorb commented:
As of subplots=True, I've only used it with histograms and line plots. I'm not familiar with their usage with scatter plots/hexbin plots. So I'm not sure how to test them. Simply using it with single plot doesn't interfere with the outcome, but I'm assuming that's not what you were referring to. I think we'll need your expertise here.

`df.plot.scatter('A label','B label',c='C label', PATCH_MODE_FLAG = True, subplots=True);`
![image](https://user-images.githubusercontent.com/26352146/37774409-7f994c9a-2db6-11e8-804b-ddfaadea46e5.png)

javadnoorb committed:
removed whitespaces

pep8speaks commented:
Hello @javadnoorb! Thanks for updating the PR.

 - In the file [`pandas/plotting/_core.py`](https://github.com/pandas-dev/pandas/blob/8f52eab88b1c2d676460a88dd2a93b74576ba50a/pandas/plotting/_core.py), following are the PEP8 issues :

> [Line 907:1](https://github.com/pandas-dev/pandas/blob/8f52eab88b1c2d676460a88dd2a93b74576ba50a/pandas/plotting/_core.py#L907): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 2798:1](https://github.com/pandas-dev/pandas/blob/8f52eab88b1c2d676460a88dd2a93b74576ba50a/pandas/plotting/_core.py#L2798): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 3216:43](https://github.com/pandas-dev/pandas/blob/8f52eab88b1c2d676460a88dd2a93b74576ba50a/pandas/plotting/_core.py#L3216): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 3216:59](https://github.com/pandas-dev/pandas/blob/8f52eab88b1c2d676460a88dd2a93b74576ba50a/pandas/plotting/_core.py#L3216): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 3216:61](https://github.com/pandas-dev/pandas/blob/8f52eab88b1c2d676460a88dd2a93b74576ba50a/pandas/plotting/_core.py#L3216): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 3295:1](https://github.com/pandas-dev/pandas/blob/8f52eab88b1c2d676460a88dd2a93b74576ba50a/pandas/plotting/_core.py#L3295): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 3301:80](https://github.com/pandas-dev/pandas/blob/8f52eab88b1c2d676460a88dd2a93b74576ba50a/pandas/plotting/_core.py#L3301): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 3301:90](https://github.com/pandas-dev/pandas/blob/8f52eab88b1c2d676460a88dd2a93b74576ba50a/pandas/plotting/_core.py#L3301): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 3301:92](https://github.com/pandas-dev/pandas/blob/8f52eab88b1c2d676460a88dd2a93b74576ba50a/pandas/plotting/_core.py#L3301): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals

 - In the file [`pandas/plotting/_tools.py`](https://github.com/pandas-dev/pandas/blob/8f52eab88b1c2d676460a88dd2a93b74576ba50a/pandas/plotting/_tools.py), following are the PEP8 issues :

> [Line 314:80](https://github.com/pandas-dev/pandas/blob/8f52eab88b1c2d676460a88dd2a93b74576ba50a/pandas/plotting/_tools.py#L314): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 79 characters)

javadnoorb reviewed (commented):
removed the whitespaces

TomAugspurger commented:
Don't worry about subplots. It would be something like `df.plot.scatter(x=0, y=[1, 2], subplots=True)`, which currently isn't allowed.

masongallo committed:
API & BUG: allow list-like y argument to df.plot & fix integer arg to x,y (#20000)

* Add support for list-like y argument

* update whatsnew

* add doc change for y

* Add test cases and fix position args

* don't copy save cols ahead of time and update whatsnew

* address fdbck

javadnoorb commented:
That test with `assert` doesn't deal with the issue. It seems like if you make a subplot and pass the axis to pandas, the x-axis comes out fine. I didn't use environments, and don't know how to switch from my development pandas to the native version without an uninstall. But with the `PATCH_MODE_FLAG=False` you can see it:

```
fig, ax1 = pl.subplots()
df.plot.scatter(x='A label', y='B label', c='C label',ax=ax1, PATCH_MODE_FLAG=False)
```
![image](https://user-images.githubusercontent.com/26352146/37789044-48531978-2dd9-11e8-85d0-4562a26ca296.png)


`df.plot.scatter(x='A label', y='B label', c='C label', PATCH_MODE_FLAG=False)`
![image](https://user-images.githubusercontent.com/26352146/37788928-0c47e0bc-2dd9-11e8-997b-50028f9edc15.png)

javadnoorb commented:
This works:

```
PATCH_MODE_FLAG = False
ax1 = df.plot.scatter(x='A label', y='B label', PATCH_MODE_FLAG=PATCH_MODE_FLAG)
ax2 = df.plot.scatter(x='A label', y='B label', c='C label', PATCH_MODE_FLAG=PATCH_MODE_FLAG)
assert (ax1.get_xticks()==ax2.get_xticks()).all(), "axis labels are not the same"
```

but regardless of `PATCH_MODE_FLAG` value it doesn't throw an error. 

I think the issue is not with the xticks attribute but rather showing them.

javadnoorb commented:
Trying a few of these. Quite bizarre:

```
PATCH_MODE_FLAG = False
ax1 = df.plot.scatter(x='A label', y='B label', PATCH_MODE_FLAG=PATCH_MODE_FLAG)
ax2 = df.plot.scatter(x='A label', y='B label', c='C label', PATCH_MODE_FLAG=PATCH_MODE_FLAG)
assert (ax1.get_xticks()==ax2.get_xticks()).all(), "Axis xticks are not the same"
assert ax1.xaxis.get_visible()==ax2.xaxis.get_visible(), "x-axis visbilities are not the same"
print(ax1.get_xticklabels(),ax2.get_xticklabels())
print(ax1.get_xticks(),ax2.get_xticks())
print(ax1.xaxis.get_visible(),ax2.xaxis.get_visible())
```

![image](https://user-images.githubusercontent.com/26352146/37790413-d5a3cbf8-2ddc-11e8-9dfd-30c1e97b6724.png)

WillAyd committed:
Removed dead groupby code (#20457)

minggli committed:
EHN: allow zip compression in `to_pickle`, `to_json`, `to_csv` (#20394)

TomAugspurger committed:
ENH: Sorting of ExtensionArrays (#19957)

igorcadelima committed:
DOC: Improve the docstring of DataFrame.transpose() (#20254)

* DOC: Improve the docstring of DataFrame.transpose()

Co-authored-by: Carlos Eduardo Moreira dos Santos <cems@cemshost.com.br>

thismakessand committed:
DOC: Update pandas.Series.copy docstring (#20261)

4kxz committed:
DOC: update the pandas.DataFrame.plot.box docstring (#20373)

Imanflow committed:
DOC: make deprecation warning more visible with red box (#20357)

jreback committed:
TST: clean deprecation warnings & some parametrizing (#20467)

TomAugspurger committed:
REF: Mock all S3 Tests (#20409)

* REF: Mock all S3 Tests

Closes https://github.com/pandas-dev/pandas/issues/19825

minggli committed:
DOC: docstring to series.unique (#20474)

TomAugspurger committed:
TST: Fixed version comparison (#20469)

* TST: Fixed version comparison

This failed to skip for 3.5.x because the micro component made it False.

* Use pandas.compat

* More pandas compat

jreback committed:
TST: 32-bit compat for categorical factorization tests (#20482)

reidy-p committed:
API: Preserve int columns in to_dict('index') (#20444)

tv3141 committed:
DOC: Fix broken dependency links (#20471)

* Change bottleneck link to github repository page.

An issue exists for the broken documentation link on pypy https://github.com/kwgoodman/bottleneck/issues/173

* Fix link to fastparquet documentation.

dcreekp committed:
DOC: update the pandas.Series.str.startswith docstring (#20458)

* DOC: update the pandas.Series.str.startswith docstring

* DOC: update the pandas.Series.str.startswith docstring

* DOC: update the pandas.Series.str.startswith docstring 2

* DOC: update the pandas.Series.str.startswith docstring 3

* DOC: update the pandas.Series.str.startswith docstring 4

JQGoh committed:
BUG: dropna() on single column timezone-aware values (#13407) (#20422)

jreback committed:
TST: test_nanops some parametrize & catch warnings (RuntimeWarning: All-Nan slice in tests) (#20484)

* TST: test_nanops some parametrize & catch warnings (RuntimeWarning: All-NaN slice in tests)

* COMPAT: work around deprecation warning on non-equal dtype comparisons

closes #20011

* WARN: bincount minlength deprecation warning

ibrahimsharaf committed:
ENH: DataFrame.pivot accepts a list of values (#18636)

jorisvandenbossche committed:
BUG: raise error when setting cached properties (#20487)

samuelsinayoko committed:
DOC: update the DataFrame.stack docstring (#20430)

TomAugspurger committed:
CI: Fixed deprecationWarning (#20489)

Closes #20479

Cheukting committed:
DOC: update the isna, isnull, notna and notnull docstring (#20459)

jorisvandenbossche committed:
CLN: remove deprecated infer_dst keyword (#20490)

TomAugspurger committed:
Parametrized NA sentinel for factorize (#20473)

GGordonGordon committed:
DEPR: deprecate get_ftype_counts (GH18243) (#20404)

fdroessler committed:
DOC: update the Series.str.join docstring (#20463)

jorisvandenbossche committed:
DOC: change to numpydoc: use isdatadescriptor instead of isgetsetdescriptor to check attributes

This is needed to ensure .columns / .index / cached properties are recognized as an attribute

TomAugspurger committed:
ENH/API: ExtensionArray.factorize (#20361)

TomAugspurger committed:
DOC: add guide on shared docstrings (#20016)

cbrnr committed:
Set pd.options.display.max_columns=0 by default (#17023)

Change `max_columns` to `0` (automatically adapt the number of displayed columns to the actual terminal width)

jreback committed:
DOC: whatsnew edits

TomAugspurger committed:
ENH: Support ExtensionArray in Groupby (#20502)

dcreekp committed:
DOC: update the pandas.Series.str.endswith docstring (#20491)

Hamishpk committed:
DOC: Improving the docstring of Series.str.upper and related (#20462)

TomAugspurger commented:
@javadnoorb how's this going? For this PR I would ignore user-provided axes, and only worry about the common case of `df.plot(x='x', y='y', c='c')`.

mrw34 committed:
DOC: update the Series.between docstring (#20443)

javadnoorb commented:
@TomAugspurger I haven't worked on this for a few days. Let me see if I can come up with a reasonable test. `get_xticklabels` returns an object, so I think we can't use it for comparison. I think the reason the x-labels disappear should be because of these two lines:

https://github.com/pandas-dev/pandas/blob/8a5830305d2d9a087fbe46bd968218104ffdfc49/pandas/plotting/_tools.py#L282
https://github.com/pandas-dev/pandas/blob/8a5830305d2d9a087fbe46bd968218104ffdfc49/pandas/plotting/_tools.py#L296

my guess is that something like:

`assert ax1.get_majorticklabels().get_visible()==ax2.get_majorticklabels().get_visible()` 

and 

`assert ax1.get_label().get_visible()==ax2.get_label().get_visible()` 

or something like that, should work for the test. 

Let me give it a shot.

TomAugspurger commented:
Sounds good. Let me know if that doesn't work and I can take a look.

On Wed, Mar 28, 2018 at 11:48 AM, Javad Noorbakhsh <notifications@github.com
> wrote:

> @TomAugspurger <https://github.com/TomAugspurger> I haven't worked on
> this for a few days. Let me see if I can come up with a reasonable test.
> get_xticklabels returns an object, so I think we can't use it for
> comparison. I think the reason the x-labels disappear should be because of
> these two lines:
>
> https://github.com/pandas-dev/pandas/blob/8a5830305d2d9a087fbe46bd968218
> 104ffdfc49/pandas/plotting/_tools.py#L282
> https://github.com/pandas-dev/pandas/blob/8a5830305d2d9a087fbe46bd968218
> 104ffdfc49/pandas/plotting/_tools.py#L296
>
> my guess is that something like:
>
> assert ax1.get_majorticklabels().get_visible()==ax2.get_
> majorticklabels().get_visible()
>
> and
>
> assert ax1.get_label().get_visible()==ax2.get_label().get_visible()
>
> or something like that, should work for the test.
>
> Let me give it a shot.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/20446#issuecomment-376955988>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIkLc6e7wel8eSgSbmlptCfR_HWFaks5ti77NgaJpZM4S2XhP>
> .
>

javadnoorb commented:
Ok, I wrote a test that seems to work:
```
def test_if_colorbar_affects_xaxis_visibility(PATCH_MODE_FLAG=True):
    random_array = np.random.random((1000,3))
    df = pd.DataFrame(random_array,columns=['A label','B label','C label'])

    ax1 = df.plot.scatter(x='A label', y='B label', PATCH_MODE_FLAG=PATCH_MODE_FLAG)
    ax2 = df.plot.scatter(x='A label', y='B label', c='C label', PATCH_MODE_FLAG=PATCH_MODE_FLAG)

    assert all([vis[0].get_visible()==vis[1].get_visible() for vis in zip(ax1.xaxis.get_minorticklabels(),ax2.xaxis.get_minorticklabels())]), 'minor x-axis tick labels visibility changes when colorbar included'
    assert all([vis[0].get_visible()==vis[1].get_visible() for vis in zip(ax1.xaxis.get_majorticklabels(),ax2.xaxis.get_majorticklabels())]), 'major x-axis tick labels visibility changes when colorbar included'
    assert ax1.xaxis.get_label().get_visible()==ax2.xaxis.get_label().get_visible(),'x-axis label visibility changes when colorbar included'
```
```    
test_if_colorbar_affects_xaxis_visibility(PATCH_MODE_FLAG=True)    
```
![image](https://user-images.githubusercontent.com/26352146/38047415-6f046c82-3290-11e8-9f00-4d7f504cf6d4.png)
![image](https://user-images.githubusercontent.com/26352146/38047422-75e82192-3290-11e8-807c-3679d298e2d7.png)
```
test_if_colorbar_affects_xaxis_visibility(PATCH_MODE_FLAG=False)
```

```---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-47-0cc25b4e00f4> in <module>()
     10     assert ax1.xaxis.get_label().get_visible()==ax2.xaxis.get_label().get_visible(),'x-axis label visibility changes when colorbar included'
     11 
---> 12 test_if_colorbar_affects_xaxis_visibility(PATCH_MODE_FLAG=False)

<ipython-input-47-0cc25b4e00f4> in test_if_colorbar_affects_xaxis_visibility(PATCH_MODE_FLAG)
      7 
      8     assert all([vis[0].get_visible()==vis[1].get_visible() for vis in zip(ax1.xaxis.get_minorticklabels(),ax2.xaxis.get_minorticklabels())]), 'minor x-axis tick labels visibility changes when colorbar included'
----> 9     assert all([vis[0].get_visible()==vis[1].get_visible() for vis in zip(ax1.xaxis.get_majorticklabels(),ax2.xaxis.get_majorticklabels())]), 'major x-axis tick labels visibility changes when colorbar included'
     10     assert ax1.xaxis.get_label().get_visible()==ax2.xaxis.get_label().get_visible(),'x-axis label visibility changes when colorbar included'
     11 

AssertionError: major x-axis tick labels visibility changes when colorbar included
```

![image](https://user-images.githubusercontent.com/26352146/38047483-afb01362-3290-11e8-847f-5c05308c5637.png)
![image](https://user-images.githubusercontent.com/26352146/38047488-b2f6755c-3290-11e8-811c-03a729cf3f08.png)


If that seems fine I'll include the changes and remove the PATCH_MODE flag. 

I'm not familiar with writing tests. Is this where it goes?
https://github.com/pandas-dev/pandas/blob/master/pandas/tests/plotting/test_misc.py

TomAugspurger commented:
That test seems sensible.

Tests can go in `pandas/pandas/tests/plotting/test_frame.py` under TestFrame. There are some other ones there for `scatter`.

javadnoorb committed:
Update _core.py

removing PATH_MODE debugging flag

javadnoorb commented:
Ok, great. I'll check if it passes the regular checks before including the test.

javadnoorb committed:
Update _core.py

remove whitespaces

javadnoorb commented:
Also need a test for hexbin. This one seems to work:

```
def test_if_hexbin_xaxis_label_is_visible(PATCH_MODE_FLAG=True):
    random_array = np.random.random((1000,3))
    df = pd.DataFrame(random_array,columns=['A label','B label','C label'])

    ax = df.plot.hexbin('A label','B label', gridsize=12, PATCH_MODE_FLAG = PATCH_MODE_FLAG);    
    assert all([vis.get_visible() for vis in ax.xaxis.get_minorticklabels()]), 'minor x-axis tick labels are not visible'
    assert all([vis.get_visible() for vis in ax.xaxis.get_majorticklabels()]), 'major x-axis tick labels are not visible'
    assert ax.xaxis.get_label().get_visible(),'x-axis label is not visible'
```

```
test_if_hexbin_xaxis_label_is_visible(PATCH_MODE_FLAG=True)   
```
![image](https://user-images.githubusercontent.com/26352146/38051264-c750eea0-329b-11e8-9bae-856e60e094c7.png)

```
test_if_hexbin_xaxis_label_is_visible(PATCH_MODE_FLAG=False)   
```
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-66-cd06ece40a4d> in <module>()
      9 
     10 
---> 11 test_if_hexbin_xaxis_label_is_visible(PATCH_MODE_FLAG=False)

<ipython-input-66-cd06ece40a4d> in test_if_hexbin_xaxis_label_is_visible(PATCH_MODE_FLAG)
      4 
      5     ax = df.plot.hexbin('A label','B label', gridsize=12, PATCH_MODE_FLAG = PATCH_MODE_FLAG);
----> 6     assert all([vis.get_visible() for vis in ax.xaxis.get_minorticklabels()]), 'minor x-axis tick labels are not visible'
      7     assert all([vis.get_visible() for vis in ax.xaxis.get_majorticklabels()]), 'major x-axis tick labels are not visible'
      8     assert ax.xaxis.get_label().get_visible(),'x-axis label is not visible'

AssertionError: minor x-axis tick labels are not visible
```
![image](https://user-images.githubusercontent.com/26352146/38051297-dc3f9d70-329b-11e8-9e8e-01fb6b249fd6.png)

javadnoorb commented:
These changes seem pretty benign. Is there any specific reason for Travis-CI to throw an error?

TomAugspurger commented on pandas/plotting/_tools.py:
```diff
@@ -311,7 +311,7 @@ def _handle_shared_axes(axarr, nplots, naxes, nrows, ncols, sharex, sharey):
                     # other off layout handles the case that the subplot is
                     # the last in the column, because below is no subplot/gap.
-                    if not layout[ax.rowNum + 1, ax.colNum]:
+                    if not layout[ax.rowNum + 1, ax.colNum] or getattr(ax, '_pandas_colorbar_axes', False):
```
This line is too long: https://travis-ci.org/pandas-dev/pandas/jobs/359516939#L3006

javadnoorb replied:
Thanks. I'll fix it. Sorry I'm pretty new to the open-source world.

mmngreco committed:
CLN: Removed not necessary bn switch decorator on nansum (#20481)

TomAugspurger committed:
DOC: Fix various warnings (#20509)

javadnoorb committed:
Update _tools.py

javadnoorb committed:
included tests for scatterplot and hexbin plot to ensure colorbar does not exclude x-axis label and tick-labels

javadnoorb committed:
Update test_frame.py

remove whitespace

jreback committed:
DOC: whatsnew edits

javadnoorb committed:
Update test_frame.py

changing for PEP8

javadnoorb committed:
Update _tools.py

changing for PEP8

javadnoorb committed:
fixing assert message

TomAugspurger commented on pandas/plotting/_core.py:
```diff
@@ -870,6 +870,7 @@ def _make_plot(self):
         scatter = ax.scatter(data[x].values, data[y].values, c=c_values,
                              label=label, cmap=cmap, **self.kwds)
         if cb:
+            ax._pandas_colorbar_axes = True
```
Add a comment here explaining why you're adding this attribute.

TomAugspurger commented on pandas/plotting/_core.py:
```diff
@@ -915,6 +917,7 @@ def _make_plot(self):
         ax.hexbin(data[x].values, data[y].values, C=c_values, cmap=cmap,
                   **self.kwds)
         if cb:
+            ax._pandas_colorbar_axes = True
```
Comment.

TomAugspurger commented on pandas/plotting/_tools.py:
```diff
@@ -311,7 +311,8 @@ def _handle_shared_axes(axarr, nplots, naxes, nrows, ncols, sharex, sharey):
                     # other off layout handles the case that the subplot is
                     # the last in the column, because below is no subplot/gap.
-                    if not layout[ax.rowNum + 1, ax.colNum]:
+                    if not layout[ax.rowNum + 1, ax.colNum] or \
```
We use parenthesis to break long lines, not `\`.

TomAugspurger commented on pandas/tests/plotting/test_frame.py:
```diff
@@ -1030,6 +1030,43 @@ def test_plot_scatter(self):
 
+    @pytest.mark.slow
+    def test_if_scatterplot_colorbar_affects_xaxis_visibility(self):
+        random_array = np.random.random((1000,3))
```
PEP8.

You can run `flake8 pandas/tests/plotting/test_frame.py` locally to check this.

TomAugspurger replied:
Include a reference to the github issue as a comment.

TomAugspurger commented on pandas/tests/plotting/test_frame.py:
```diff
@@ -1030,6 +1030,43 @@ def test_plot_scatter(self):
+        ax1 = df.plot.scatter(x='A label', y='B label')
+        ax2 = df.plot.scatter(x='A label', y='B label', c='C label')
+
+        assert all([vis[0].get_visible() == vis[1].get_visible() for vis in
```
Just do the `assert`, no error message. pytest will ensure that if the test fails, we still get a meaningful error message.

TomAugspurger replied:
Further, I think you can just do

```python
vis1 = [vis.get_visible() for vis in ax1.xaxis.get_minorticklabels()]
vis2 = [vis.get_visible() for vis in ax2.xaxis.get_minorticklabels()]

assert vis1 == vis2
```

As written, if one happens to be shorter than the other, we wouldn't know because the `zip` would just stop.

TomAugspurger reviewed (commented):
Good start. Could you include a release note under the bugfix section (I think we have a vis subsection). 0.23.0.txt.

javadnoorb committed:
Update test_frame.py

fixing lint E231

kylebarron committed:
DOC: Add comparison with Stata page to documentation (#19945)

* Add Comparison with Stata to documentation

* Add comparison_with_stata.rst to index.rst.template

* Fix section underlining

* Small fixes

pulkitmaloo committed:
DOC: Update missing_data.rst (#20424)

javadnoorb committed:
fixing style issues

TomAugspurger commented:
https://github.com/pandas-dev/pandas/blob/master/doc/source/whatsnew/v0.23.0.txt

On Thu, Mar 29, 2018 at 1:23 PM, Javad Noorbakhsh <notifications@github.com>
wrote:

> Where are the release notes?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/20446#issuecomment-377327656>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIhj82OsukyNfHE-kQT8ohbcDzhN1ks5tjSaHgaJpZM4S2XhP>
> .
>

javadnoorb committed:
Update v0.23.0.txt

plotting bugfixes

javadnoorb committed:
Merge branch 'master' into master

javadnoorb committed:
Update test_frame.py

extra comments pointing to issues

javadnoorb committed:
remove whitespace

javadnoorb commented:
I've been trying the tests with the inline and pdf backend. Tests pass depending on the backend. So I'm guessing that my PR doesn't really address the issue.

TomAugspurger commented:
> Tests pass depending on the backend. So I'm guessing that my PR doesn't really address the issue.

Can you rephrase that? This behavior only manifests with specific backends. Ideally it would always fail on a given backend (e.g. notebook) without your changes, and always pass with it.

If the backend is too much of an issue, we can adapt the test. The change is clearly an improvement.

javadnoorb committed:
Inline backend (#2)

* minor change in comment

* added condition on inline backend

* added condition on inline backend

* issue with floats determining the position of axes corners in ipython

javadnoorb committed:
Update _tools.py

javadnoorb committed:
Update _core.py

javadnoorb committed:
Update _core.py

javadnoorb committed:
Update v0.23.0.txt

javadnoorb committed:
Update _core.py

javadnoorb commented:
@TomAugspurger. Seems like the issue is because in the inline backend floats corresponding to location of axes have too many arbitrary trailing digits. For example two axes that have their bottom left corners at 0.3 may appear as 0.3 and 0.30000001. I rounded them in `_core.py`.

This new commit works for single plots with pdf and inline backends. It still fails for multiplots. We'll need to dig deeper for that. But it seems like we don't need the previous hack anymore.

TomAugspurger commented on pandas/plotting/_core.py:
```diff
@@ -788,6 +788,11 @@ def _get_axes_layout(self):
+            #  `get_points()` have too many ad hoc trailing
+            # digits. Unless rounded these values won't
+            # match in the following set operations
+            points = np.round(points, 3)
```
I worry about this causing debugging nightmares further down the road... I'd prefer to solve this local to the problem if possible.

If not, you could perhaps include a keyword to `_get_axes_layout`? I'm not sure how difficult that would be to plumb through.

javadnoorb replied:
I agree. That's pretty tough to track. 

I made a new commit. The changes are local now. I'm forcing colorbars to take on their relevant axes positions from the parent axes. The merge passes the tests and resolves the xlabel issue (#10611 and #10678) but not with the subplots (#20455).

javadnoorb committed:
Inline backend - fixing colorbar axis position roundoff error locally

* minor change in comment

* added condition on inline backend

* added condition on inline backend

* issue with floats determining the position of axes corners in ipython

* fixing axis position roundoff error locally

javadnoorb committed:
broken merge 

* minor change in comment

* added condition on inline backend

* added condition on inline backend

* issue with floats determining the position of axes corners in ipython

* fixing axis position roundoff error locally

* broken merge

javadnoorb commented:
I don't know what happened with Travis-CI, but looks like there was a connection problem in the middle of the run. I tried this on my own fork and Travis only threw two errors instead (pasted below). It looks like `read_html` is trying to read a dead link for testing:

http://ndb.nal.usda.gov/ndb/foods/show/1732?fg=&man=&lfacet=&format=&count=&max=25&offset=&sort=&qlookup=spam

Any ideas what happened?


```
=================================== FAILURES ===================================
_______________________ TestReadHtml.test_spam_url[bs4] ________________________
[gw1] linux2 -- Python 2.7.14 /home/travis/miniconda3/envs/pandas/bin/python
self = <pandas.tests.io.test_html.TestReadHtml object at 0x7f2989916310>
    @network
    def test_spam_url(self):
        url = ('http://ndb.nal.usda.gov/ndb/foods/show/1732?fg=&man=&'
               'lfacet=&format=&count=&max=25&offset=&sort=&qlookup=spam')
>       df1 = self.read_html(url, '.*Water.*')
pandas/tests/io/test_html.py:106: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/io/html.py:983: in read_html
    displayed_only=displayed_only)
pandas/io/html.py:811: in _parse
    raise_with_traceback(retained)
pandas/io/html.py:793: in _parse
    tables = p.parse_tables()
pandas/io/html.py:213: in parse_tables
    tables = self._parse_tables(self._build_doc(), self.match, self.attrs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <pandas.io.html._BeautifulSoupHtml5LibFrameParser object at 0x7f2989598310>
doc = <!DOCTYPE html>\n<!--[if lt IE 7 ]> <html lang="en" class="no-js ie6"> <![endi...t\t</div>\n\t\t</small>\n\t\t</footer>\n\t\t\n\t\t\n    \n   \n\n</body></html>
match = <_sre.SRE_Pattern object at 0x7f29899c2b20>, attrs = None
    def _parse_tables(self, doc, match, attrs):
        element_name = self._strainer.name
        tables = doc.find_all(element_name, attrs=attrs)
    
        if not tables:
>           raise ValueError('No tables found')
E           ValueError: No tables found
pandas/io/html.py:467: ValueError
_______________________ TestReadHtml.test_spam_url[lxml] _______________________
[gw1] linux2 -- Python 2.7.14 /home/travis/miniconda3/envs/pandas/bin/python
self = <pandas.tests.io.test_html.TestReadHtml object at 0x7f298ca9fe10>
    @network
    def test_spam_url(self):
        url = ('http://ndb.nal.usda.gov/ndb/foods/show/1732?fg=&man=&'
               'lfacet=&format=&count=&max=25&offset=&sort=&qlookup=spam')
>       df1 = self.read_html(url, '.*Water.*')
pandas/tests/io/test_html.py:106: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/io/html.py:983: in read_html
    displayed_only=displayed_only)
pandas/io/html.py:811: in _parse
    raise_with_traceback(retained)
pandas/io/html.py:793: in _parse
    tables = p.parse_tables()
pandas/io/html.py:213: in parse_tables
    tables = self._parse_tables(self._build_doc(), self.match, self.attrs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <pandas.io.html._LxmlFrameParser object at 0x7f298ca9f610>
doc = <Element html at 0x7f298c980730>
match = <_sre.SRE_Pattern object at 0x7f29899c2b20>, kwargs = None
    def _parse_tables(self, doc, match, kwargs):
        pattern = match.pattern
    
        # 1. check all descendants for the given pattern and only search tables
        # 2. go up the tree until we find a table
        query = '//table//*[re:test(text(), {patt!r})]/ancestor::table'
        xpath_expr = u(query).format(patt=pattern)
    
        # if any table attributes were given build an xpath expression to
        # search for them
        if kwargs:
            xpath_expr += _build_xpath_expr(kwargs)
    
        tables = doc.xpath(xpath_expr, namespaces=_re_namespace)
    
        tables = self._handle_hidden_tables(tables, "attrib")
        if self.displayed_only:
            for table in tables:
                # lxml utilizes XPATH 1.0 which does not have regex
                # support. As a result, we find all elements with a style
                # attribute and iterate them to check for display:none
                for elem in table.xpath('.//*[@style]'):
                    if "display:none" in elem.attrib.get(
                            "style", "").replace(" ", ""):
                        elem.getparent().remove(elem)
    
        if not tables:
            raise ValueError("No tables found matching regex {patt!r}"
>                            .format(patt=pattern))
E           ValueError: No tables found matching regex '.*Water.*'
pandas/io/html.py:586: ValueError
```

javadnoorb committed:
Merge remote-tracking branch 'upstream/master'

javadnoorb committed:
Merge remote-tracking branch 'upstream/master'

javadnoorb commented:
I just remembered about this. The last commit seemed to have passed all the checks. Is there any interest to follow up on it?

TomAugspurger commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -1176,7 +1176,7 @@ Plotting
 - :meth:`Series.plot.kde` has exposed the args ``ind`` and ``bw_method`` in the docstring (:issue:`18461`). The argument ``ind`` may now also be an integer (number of sample points).
 - :func:`DataFrame.plot` now supports multiple columns to the ``y`` argument (:issue:`19699`)
-
+- Bug in :func:'DataFrame.plot.scatter' and :func:'DataFrame.plot.hexbin' caused x-axis label and ticklabels to disappear when colorbar was on in IPython inline backend (:issue:`10611` and :issue:`10678`)
```
This should be moved to 0.24.0

TomAugspurger commented on pandas/plotting/_core.py:
```diff
@@ -882,8 +882,13 @@ def _make_plot(self):
                 kws['label'] = c if c_is_column else ''
-            self.fig.colorbar(img, **kws)
-
+            cbar = self.fig.colorbar(img, **kws)
```
Maybe refactor this and https://github.com/pandas-dev/pandas/pull/20446/files#diff-2b118bda866f3d626ceb6b529c62cd1aR932 to a method. They're identical, right? I guess aside from `**kws` vs. `ax=ax`?

TomAugspurger reviewed (commented):
Sorry for dropping this. Can you merge in master and move the whatsnew to 0.24.0?

Looks good otherwise, so ping on green.

javadnoorb committed:
update fork

Merge branch 'master' of https://github.com/pandas-dev/pandas

javadnoorb committed:
refactoring

javadnoorb committed:
refactoring

javadnoorb committed:
updating whatsnew docs

javadnoorb committed:
updating whatsnew docs

javadnoorb committed:
updating whatsnew docs

javadnoorb committed:
updating whatsnew docs

jreback commented on pandas/plotting/_core.py:
```diff
@@ -833,6 +833,23 @@ def _post_plot_logic(self, ax, data):
         ax.set_ylabel(pprint_thing(y))
         ax.set_xlabel(pprint_thing(x))
 
+    def _plot_colorbar(self, kws):
```
can you make ax a named arg here and so you would pass as ``**kwds`` when you call.

jreback commented on pandas/plotting/_core.py:
```diff
@@ -833,6 +833,23 @@ def _post_plot_logic(self, ax, data):
+        # To deal with this, this method forces the colorbar
+        # height to take the height of the parent axes.
+        ax = kws['ax']
+        img = ax.collections[0]
```
is there an issue we can reference about this impl. IOW next person to look at wouldn't know where these set_position come from. so a) ref this issue in a comment and b) an Ipython refernce (if you can)

jreback reviewed (changes requested):
comments, otherwise lgtm. ping on green.

javadnoorb committed:
added more comments

jreback commented:
thanks @javadnoorb very nice!

javadnoorb commented:
thanks @jreback!
